Implemented various kick_with_reason() methods.

vicky5124 opened the pull request:
modified:   src/http/client.rs
	modified:   src/http/routing.rs
	modified:   src/model/guild/guild_id.rs
	modified:   src/model/guild/member.rs
	modified:   src/model/guild/mod.rs
	modified:   src/model/guild/partial_guild.rs
	modified:   src/utils/mod.rs

Implemented a new method `model::guild::Member.kick_with_reason(Http, Option<&str>)` that allows the user to kick a member from a guild with an optional reason. 
Should not be a breaking change. (EDIT: it is a breaking change due to the kick_member() method)

vicky5124 committed:
Implemented method kick_with_reason() to model::guild::Member
	modified:   src/http/client.rs
	modified:   src/http/routing.rs
	modified:   src/model/guild/guild_id.rs
	modified:   src/model/guild/member.rs
	modified:   src/model/guild/mod.rs
	modified:   src/model/guild/partial_guild.rs
	modified:   src/utils/mod.rs

arqunis commented on src/http/client.rs:
```diff
@@ -1358,11 +1358,22 @@ impl Http {
+            r = reason
+        } else {
+            r = Some("");
+        }
```
```suggestion
```

arqunis commented on src/http/client.rs:
```diff
@@ -1358,11 +1358,22 @@ impl Http {
+            route: RouteInfo::KickMember {
+                guild_id,
+                user_id,
+                reason: r
```
```suggestion
                // Supplement an empty string when `reason` is `None`.
                reason: reason.or(Some("")),
```

vicky5124 committed:
Update src/http/client.rs

Co-Authored-By: Alex M. M. <acdenissk69@gmail.com>

vicky5124 committed:
Update src/http/client.rs

Co-Authored-By: Alex M. M. <acdenissk69@gmail.com>

NovusTheory commented on src/model/guild/member.rs:
```diff
@@ -325,6 +325,37 @@ impl Member {
+    ///
+    /// [`kick`]: #method.kick
+    #[cfg(feature = "http")]
+    pub fn kick_with_reason(&self, cache_http: impl CacheHttp, reason: Option<&str>) -> Result<()> {
```
I don't know if it's just me but it feels weird to have a method named kick_with_reason but making the reason optional. It kind of seems to counteract what the method name is wanting.

vicky5124 replied:
Did this so Args.remains() could be used directly as a reason

arqunis replied:
I agree with Novus. If the method tells its purpose is the ability to kick someone from a guild with a reason as an explanation why, it should be a given that it will always expect a reason from the user. It should never be `None`.

This should apply to the other `kick`s you have changed. I would like to relieve the library from breaking the interface too much for something as simple as kicking with a reason.

Scotsguy commented on src/http/client.rs:
```diff
@@ -1358,11 +1358,17 @@ impl Http {
 
     /// Kicks a member from a guild.
-    pub fn kick_member(&self, guild_id: u64, user_id: u64) -> Result<()> {
+    pub fn kick_member(&self, guild_id: u64, user_id: u64, reason: Option<&str>) -> Result<()> {
```
This is a `pub` function, so this is a breaking change, is it not?

vicky5124 replied:
It seems so.
Easily fixed by passing None as an additional argument

Scotsguy replied:
Easily fixed, sure, but a breaking change nonetheless

vicky5124 committed:
Fixed missmatched branches
	modified:   src/utils/mod.rs

vicky5124 committed:
Removed Option, replaced with `&str`
	modified:   src/http/client.rs
	modified:   src/http/routing.rs
	modified:   src/model/guild/guild_id.rs
	modified:   src/model/guild/member.rs
	modified:   src/model/guild/mod.rs
	modified:   src/model/guild/partial_guild.rs

vicky5124 committed:
Updated example to follow recent changes.
	modified:   src/model/guild/member.rs

arqunis commented on src/http/client.rs:
```diff
@@ -1367,7 +1367,7 @@ impl Http {
                 user_id,
                 // Supplement an empty string when `reason` is `None`.
-                reason: reason.or(Some("")),
+                reason: reason,
```
```suggestion
                reason,
```

arqunis commented on src/http/client.rs:
```diff
@@ -1367,7 +1367,7 @@ impl Http {
                 guild_id,
                 user_id,
                 // Supplement an empty string when `reason` is `None`.
```
```suggestion

```

arqunis commented on src/model/guild/member.rs:
```diff
@@ -355,7 +355,7 @@ impl Member {
     /// [`kick`]: #method.kick
     #[cfg(feature = "http")]
-    pub fn kick_with_reason(&self, cache_http: impl CacheHttp, reason: Option<&str>) -> Result<()> {
+    pub fn kick_with_reason(&self, cache_http: impl CacheHttp, reason: &str) -> Result<()> {
```
Alright, you have almost got my all of wishes, but you seem to have missed one from my last review. 
I wanted for you administer this name change to all of the other `kick` methods as well, so that we wouldn't break the higher-level interface people would most likely use.

What I mean is that, for each `kick` method, there should be an accompanying `kick_with_reason` that the former would use. In the same way you have done it with `Member::kick`.

vicky5124 committed:
Missed previously used comment. Update src/http/client.rs

Co-Authored-By: Alex M. M. <acdenissk69@gmail.com>

vicky5124 committed:
Update src/http/client.rs

Co-Authored-By: Alex M. M. <acdenissk69@gmail.com>

vicky5124 committed:
Removed breaking changes.
	modified:   src/http/client.rs
	modified:   src/model/guild/guild_id.rs
	modified:   src/model/guild/member.rs
	modified:   src/model/guild/mod.rs
	modified:   src/model/guild/partial_guild.rs
